### PR TITLE
implemented debug() editor function which lets you navigate in the stack frame

### PR DESCRIPTION
**Pull request by gdementen**
run_editor_on_exception now uses debug so that we can navigate the traceback

PS: I am not really happy with the names usercode_traceback (which already existed) and usercode_frame (new). 
I just couldn't find anything better. Do you have a suggestion?

### CODE REVIEW

File: larray_editor/api.py [hunk 1]
```diff
@@ -333,29 +370,31 @@ def excepthook(type, value, tback):
...
+        stack_pos = user_tb_length - 1 if user_tb_length is not None and usercode_frame else None
         print("\nlaunching larray editor to debug...", file=sys.stderr)
-        edit(_get_vars_from_frame(current_tb.tb_frame))
+        _debug(stack, stack_pos=stack_pos)
 
     return excepthook
 
 
-def run_editor_on_exception(root_path=None, usercode_traceback=True):
+def run_editor_on_exception(root_path=None, usercode_traceback=True, usercode_frame=True):
```
Comment (alixdamman) [inline larray_editor/api.py]: Don't forget to repeat the new argument in the file ``viewer/__init__.pt`` in the ``larray`` repository

File: larray_editor/api.py [hunk 2]
```diff
@@ -182,6 +186,43 @@ def view(obj=None, title='', depth=0, display_caller_info=True):
...
+    dlg = MappingEditor(parent)
+    setup_ok = dlg.setup_and_check(stack_summary, stack_pos=stack_pos)
+    if setup_ok:
+        dlg.show()
+        _app.exec_()
+
+    sys.excepthook = orig_except_hook
+
+
+def debug(depth=0):
```
Comment (alixdamman) [inline larray_editor/api.py]: add this function to the ``larray`` repository ?

File: larray_editor/editor.py [hunk 3]
```diff
@@ -394,18 +422,32 @@ def void_formatter(array, *args, **kwargs):
...
                 self.new()
-        # convert input data to Session if not
-        else:
+        elif not debug:
             self._push_data(data)
 
     def _push_data(self, data):
         self.data = data if isinstance(data, Session) else Session(data)
         if qtconsole_available:
+            # def test_user_func():
```
Comment (alixdamman) [inline larray_editor/editor.py]: remove lines in comment if they are no longer needed ?

**Comment by alixdamman** [comment]
`usersession` instead of `usercode`?

File: larray_editor/api.py (hunk 2 above)
Comment (gdementen) [reply larray_editor/api.py]: will do

File: larray_editor/api.py (hunk 1 above)
Comment (gdementen) [reply larray_editor/api.py]: will do

**Comment by gdementen** [comment]
> `usersession` instead of `usercode`?

Doesn't seem better. The "usercode" part of the names is fine IMO, but the "traceback" and "frame" parts are probably chinese to most of our users... even though this is not a big deal since they are not the target for such arguments. I expect we (or I?) will be the only one(s) using those arguments, but to me at least, they are very useful. I only want our typical users to understand the arguments enough to know they should probably not touch them 😉, and developers like you to understand them enough to actually use them.

**Comment by alixdamman** [comment]
In that case, merge the PR as it.

**Pull request closed without merging**